src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/IncentiveSummaryService.kt

```diff
@@ -15,70 +18,63 @@ class IncentiveSummaryService(
-                  PrisonerIncentiveSummary(
-                    firstName = WordUtils.capitalizeFully(p.firstName),
-                    lastName = WordUtils.capitalizeFully(p.lastName),
-                    prisonerNumber = p.offenderNo,
-                    bookingId = p.bookingId,
-                    imageId = p.facialImageId,
-                    daysOnLevel = tuples.t2[p.bookingId]?.daysOnLevel ?: 0,
-                    daysSinceLastReview = tuples.t2[p.bookingId]?.daysSinceReview ?: 0,
-                    positiveBehaviours = tuples.t3[p.offenderNo]?.totalCaseNotes ?: 0,
-                    incentiveEncouragements = tuples.t3[p.offenderNo]?.numSubTypeCount ?: 0,
-                    negativeBehaviours = tuples.t4[p.offenderNo]?.totalCaseNotes ?: 0,
-                    incentiveWarnings = tuples.t4[p.offenderNo]?.numSubTypeCount ?: 0,
-                    provenAdjudications = tuples.t5[p.bookingId]?.provenAdjudicationCount ?: 0,
-                  )
-                }.sortedWith(sortBy.applySorting(sortDirection))
+  suspend fun getIncentivesSummaryByLocation(
+    prisonId: String,
+    locationId: String,
+    sortBy: SortColumn = SortColumn.NAME,
+    sortDirection: Sort.Direction = Sort.Direction.ASC
+  ): BehaviourSummary {
+
+    val prisoners = prisonApiService.findPrisonersAtLocation(prisonId, locationId).toList()
+    val bookingIds = prisoners.map { p -> p.bookingId }.toList()
+    val offenderNos = prisoners.map { p -> p.offenderNo }.toList()
+
+    if (bookingIds.isEmpty()) throw NoPrisonersAtLocationException(prisonId, locationId)
+
+    return coroutineScope {
+
+      val positiveCaseNotes = async { getCaseNoteUsage("POS", "IEP_ENC", offenderNos) }
+      val negativeCaseNotes = async { getCaseNoteUsage("NEG", "IEP_WARN", offenderNos) }
+      val provenAdjudications = async { getProvenAdjudications(bookingIds) }
+      val iepDetails = getIEPDetails(bookingIds)
+
+      val levels = prisonApiService.getIepLevelsForPrison(prisonId).toList()
+
+      val prisonersByLevel = getPrisonersByLevel(prisoners, iepDetails)
+        .map { prisonerIepLevelMap ->
+          val iepLevel = lookupIepLevel(prisonerIepLevelMap, levels.associateBy { it.iepDescription })
+          IncentiveLevelSummary(
+            level = iepLevel.iepLevel,
+            levelDescription = iepLevel.iepDescription,
+            prisonerBehaviours = prisonerIepLevelMap.value.map { p ->
+              PrisonerIncentiveSummary(
+                firstName = WordUtils.capitalizeFully(p.firstName),
+                lastName = WordUtils.capitalizeFully(p.lastName),
+                prisonerNumber = p.offenderNo,
+                bookingId = p.bookingId,
+                imageId = p.facialImageId,
+                daysOnLevel = iepDetails[p.bookingId]?.daysOnLevel ?: 0,
+                daysSinceLastReview = iepDetails[p.bookingId]?.daysSinceReview ?: 0,
+                positiveBehaviours = positiveCaseNotes.await()[p.offenderNo]?.totalCaseNotes ?: 0,
```

Comment:
this looks interesting - not sure what happens when `await` is called multiple times on the same item?

Reply:
Seems to work